chemprop/models/ffn.py

```diff
@@ -0,0 +1,250 @@
+from typing import List, Tuple
+
+import torch
+import torch.nn as nn
+import re
+
+
+class AttrProxy(object):
+    """Translates index lookups into attribute lookups."""
+    def __init__(self, module, prefix):
+        self.module = module
+        self.prefix = prefix
+
+    def __len__(self):
+        return len([x for x in self.module.__dict__['_modules'].keys() if re.match(f'{self.prefix}\d+', x)])
+
+    def __getitem__(self, item):
+        if item >= len(self):
+            raise IndexError
+        return getattr(self.module, self.prefix + str(item))
+
+
+class MultiReadout(nn.Module):
+    """A fake list of FFNs for reading out as suggested in
+    https://discuss.pytorch.org/t/list-of-nn-module-in-a-nn-module/219/3 """
+
+    def __init__(self,
+                 features_size: int,
+                 hidden_size: int,
+                 num_layers: int,
+                 output_size: int,
+                 dropout: nn.Module,
+                 activation: nn.Module,
+                 atom_targets: List[str] = None,
+                 bond_targets: List[str] = None,
+                 atom_constraints: List[bool] = None,
+                 bond_constraints: List[bool] = None):
+        """
+        :param features_size: Dimensionality of input features.
+        :param hidden_size: Dimensionality of hidden layers.
+        :param num_layers: Number of layers in FFN.
+        :param output_size: The size of output.
+        :param dropout: Dropout probability.
+        :param activation: Activation function.
+        :param atom_targets: A list of names for atomic targets.
+        :param bond_targets: A list of names for bond targets.
+        :param atom_constraints: A list of booleans indicatin whether constraints applied to output of atomic properties.
+        :param bond_constraints: A list of booleans indicatin whether constraints applied to output of bond properties.
+        """
+        super(MultiReadout, self).__init__()
+        ind = 0
+        for i in range(len(atom_targets)):
+            constraint = atom_constraints[i]
+            self.add_module(f'readout_{ind}', FFNAtten(features_size, hidden_size, num_layers, output_size,
+                                                       dropout, activation, constraint, ffn_type='atom'))
+            ind += 1
+
+        for i in range(len(bond_targets)):
+            constraint = bond_constraints[i]
+            self.add_module(f'readout_{ind}', FFNAtten(2*features_size, hidden_size, num_layers, output_size,
+                                                       dropout, activation, constraint, ffn_type='bond'))
+            ind += 1
```

Comment:
With this loop structure, do the different atom and bond tasks all have different FFN weights?

If so, then multitasking could have the different tasks cooperate on the MPN weights but all the FFN weights are independent?